File: aat-android/src/main/java/ch/bailu/aat/gpx/GpxNodeFinder.java
package ch.bailu.aat.gpx;

import ch.bailu.aat.coordinates.BoundingBoxE6;

public class GpxNodeFinder extends GpxListWalker {

    private final BoundingBoxE6 bounding;
    private GpxPointNode node;
    private int index=0;

    public GpxNodeFinder(BoundingBoxE6 b) {
        bounding = b;
    }


    @Override
    public boolean doList(GpxList s) {
        return BoundingBoxE6.doOverlap(s.getDelta().getBoundingBox(), bounding);
    }

    @Override
    public boolean doSegment(GpxSegmentNode s) {
        if (haveNode()) {
            return false;

        } else if (BoundingBoxE6.doOverlap(s.getBoundingBox(), bounding)) {
            return true;

        } else {
            index = index + s.getSegmentSize();
            return false;
        }
    }

    @Override
    public boolean doMarker(GpxSegmentNode s) {
        return doSegment(s);
    }

    @Override
    public void doPoint(GpxPointNode point) {
        if (haveNode() == false) {
            if (bounding.contains(point)) {
                node = point;
            } else {
                index++;
            }
        }
    }

    public boolean haveNode() {
        return node != null;
    }


    public GpxPointNode getNode() {
        return node;
    }

    public int getNodeIndex() {
        return index;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/GpxNodeFinder.kt
package ch.bailu.aat_lib.gpx

import ch.bailu.aat_lib.coordinates.BoundingBoxE6
import ch.bailu.aat_lib.coordinates.BoundingBoxE6.Companion.doOverlap

/** Walks a {@link GpxList} to find the first point that falls within a given bounding box. */
class GpxNodeFinder(private val bounding: BoundingBoxE6) : GpxListWalker() {
    var node: GpxPointNode? = null
        private set
    var nodeIndex: Int = 0
        private set

    override fun doList(list: GpxList): Boolean {
        return doOverlap(list.getDelta().getBoundingBox(), bounding)
    }

    override fun doSegment(segment: GpxSegmentNode): Boolean {
        if (haveNode()) {
            return false
        } else if (doOverlap(segment.getBoundingBox(), bounding)) {
            return true
        } else {
            this.nodeIndex = this.nodeIndex + segment.segmentSize
            return false
        }
    }

    override fun doMarker(segment: GpxSegmentNode): Boolean {
        return doSegment(segment)
    }

    override fun doPoint(point: GpxPointNode) {
        if (!haveNode()) {
            if (bounding.contains(point)) {
                node = point
            } else {
                this.nodeIndex++
            }
        }
    }

    fun haveNode(): Boolean {
        return node != null
    }
}
